contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol

```diff
@@ -0,0 +1,199 @@
+    uint32 callbackGasLimit,
+    uint32 numWords,
+    address indexed sender
+  );
+  event RandomWordsFulfilled(uint256 indexed requestId, uint256 outputSeed, uint96 payment, bool success);
+  event SubscriptionCreated(uint64 indexed subId, address owner);
+  event SubscriptionFunded(uint64 indexed subId, uint256 oldBalance, uint256 newBalance);
+  event SubscriptionCanceled(uint64 indexed subId, address to, uint256 amount);
+
+  uint64 s_currentSubId;
+  uint256 s_nextRequestId = 1;
+  uint256 s_nextPreSeed = 100;
+  struct Subscription {
+    address owner;
+    uint96 balance;
+  }
+  mapping(uint64 => Subscription) s_subscriptions; /* subId */ /* subscription */
+
+  struct Request {
+    uint64 subId;
+    uint32 callbackGasLimit;
+    uint32 numWords;
+  }
+  mapping(uint256 => Request) s_requests; /* requestId */ /* request */
+
+  constructor(uint96 _baseFee, uint96 _gasPriceLink) {
+    BASE_FEE = _baseFee;
+    GAS_PRICE_LINK = _gasPriceLink;
+  }
+
+  /**
+   * @notice fulfillRandomWords fulfills the given request, sending the random words to the supplied
+   * @notice consumer.
+   *
+   * @dev This mock uses a simplified formula for calculating payment amount and gas usage, and does
+   * @dev not account for all edge cases handled in the real VRF coordinator. When making requests
+   * @dev against the real coordinator a small amount of additional LINK is required.
+   *
+   * @param _requestId the request to fulfill
+   * @param _consumer the VRF randomness consumer to send the result to
+   */
+  function fulfillRandomWords(uint256 _requestId, address _consumer) external {
+    uint256 startGas = gasleft();
+    if (s_requests[_requestId].subId == 0) {
+      revert("nonexistent request");
+    }
+    Request memory req = s_requests[_requestId];
+
+    uint256[] memory words = new uint256[](req.numWords);
+    for (uint256 i = 0; i < req.numWords; i++) {
+      words[i] = uint256(keccak256(abi.encode(_requestId, i)));
+    }
+
+    VRFConsumerBaseV2 v;
+    bytes memory callReq = abi.encodeWithSelector(v.rawFulfillRandomWords.selector, _requestId, words);
+    (bool success, ) = _consumer.call{gas: req.callbackGasLimit}(callReq);
+
+    uint96 payment = uint96(BASE_FEE + ((startGas - gasleft()) * GAS_PRICE_LINK));
+    if (s_subscriptions[req.subId].balance < payment) {
+      revert InsufficientBalance();
+    }
+    s_subscriptions[req.subId].balance -= payment;
+    delete (s_requests[_requestId]);
+    emit RandomWordsFulfilled(_requestId, _requestId, payment, success);
+  }
+
+  /**
+   * @notice fundSubscription allows funding a subscription with an arbitrary amount for testing.
+   *
+   * @param _subId the subscription to fund
+   * @param _amount the amount to fund
+   */
+  function fundSubscription(uint64 _subId, uint96 _amount) public {
+    if (s_subscriptions[_subId].owner == address(0)) {
+      revert InvalidSubscription();
+    }
+    s_subscriptions[_subId].balance = s_subscriptions[_subId].balance + _amount;
```

Comment:
I think we'd want to emit `SubscriptionFunded` here.

Reply:
Oh yeah good catch, done.